pennylane_sf/gbs.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Strawberry Fields variational GBS device
+========================================
+
+**Module name:** :mod:`pennylane_sf.gbs`
+
+.. currentmodule:: pennylane_sf.gbs
+
+The Strawberry Fields variational GBS device provides a way to encode variational parameters into
+GBS so that the gradient with respect to the output probability distribution is accessible.
+
+Classes
+-------
+
+.. autosummary::
+   StrawberryFieldsGBS
+
+Code details
+~~~~~~~~~~~~
+"""
+from collections import OrderedDict
+
+import numpy as np
+import pennylane as qml
+import strawberryfields as sf
+from pennylane.operation import Probability
+from pennylane.wires import Wires
+from strawberryfields.ops import GraphEmbed, MeasureFock
+from strawberryfields.utils import all_fock_probs_pnr
+from thewalrus.quantum import find_scaling_adjacency_matrix as rescale
+from thewalrus.quantum import photon_number_mean_vector
+
+from .expectations import identity
+from .simulator import StrawberryFieldsSimulator
+
+
+class StrawberryFieldsGBS(StrawberryFieldsSimulator):
+    r"""StrawberryFields variational GBS device for PennyLane.
+
+    This device provides a method to embed variational parameters into GBS such that the analytic
+    gradient of the probability distribution is accessible.
+
+    Args:
+        wires (int): the number of modes to initialize the device in
+        analytic (bool): indicates if the device should calculate expectations
+            and variances analytically
+        cutoff_dim (int): Fock-space truncation dimension
+        backend (str): name of the remote backend to be used
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. If ``analytic=True``,
+            this setting is ignored.
+        hbar (float): the convention chosen in the canonical commutation
+            relation :math:`[x, p] = i \hbar`
+    """
+    name = "Strawberry Fields variational GBS PennyLane plugin"
+    short_name = "strawberryfields.gbs"
+
+    _operation_map = {
+        "ParamGraphEmbed": GraphEmbed,
+    }
```

Comment:
Lol is it even worth having an operation map in this case? I suppose `StrawberryFieldsSimulator` forces you to

Reply:
I think that the appearance of `Graph` here may be a bit mysterious since we have only been discussing a matrix A. There has to be a better name, but still, how about `MatrixEmbed`?